system/include/compat/avxintrin.h

```diff
@@ -44,215 +522,1333 @@ _mm_broadcast_ss(const float *__mem_addr)
+      __ret = _mm_cmpgt_ss((__a), (__b));                                      \
+    if ((__imm) == _CMP_TRUE_UQ || (__imm) == _CMP_TRUE_US)                    \
+      __ret = _mm_move_ss((__a), (__m128)wasm_i8x16_splat(0xFF));              \
+    if ((__imm) == _CMP_NLE_US || (__imm) == _CMP_NLE_UQ)                      \
+      __ret = _mm_cmpnle_ss((__a), (__b));                                     \
+    __ret;                                                                     \
+  })
+
+static __inline__ __m256d __attribute__((__always_inline__, __nodebug__))
+_mm256_cmp_pd(__m256d a, __m256d b, const int imm8) {
+  __m256d ret;
+  ret.v0 = _mm_cmp_pd(a.v0, b.v0, imm8);
+  ret.v1 = _mm_cmp_pd(a.v1, b.v1, imm8);
+  return ret;
+}
+
+static __inline__ __m256 __attribute__((__always_inline__, __nodebug__))
+_mm256_cmp_ps(__m256 __a, __m256 __b, const int imm8) {
+  __m256 ret;
+  ret.v0 = _mm_cmp_ps(__a.v0, __b.v0, imm8);
+  ret.v1 = _mm_cmp_ps(__a.v1, __b.v1, imm8);
+  return ret;
+}
+
+#define _mm256_extract_epi32(X, N)                                             \
+  __extension__({                                                              \
+    ((N) & 0x7) < 4 ? _mm_extract_epi32((X).v0, (N) & 0x3)                     \
+                    : _mm_extract_epi32((X).v1, (N) & 0x3);                    \
+  })
+
+#define _mm256_extract_epi16(X, N)                                             \
+  __extension__({                                                              \
+    ((N) & 0xF) < 8 ? _mm_extract_epi16((X).v0, (N) & 0x7)                     \
+                    : _mm_extract_epi16((X).v1, (N) & 0x7);                    \
+  })
+
+#define _mm256_extract_epi8(X, N)                                              \
+  __extension__({                                                              \
+    ((N) & 0x1F) < 16 ? _mm_extract_epi8((X).v0, (N) & 0xF)                    \
+                      : _mm_extract_epi8((X).v1, (N) & 0xF);                   \
+  })
+
+#define _mm256_extract_epi64(X, N)                                             \
+  __extension__({                                                              \
+    ((N) & 0x3) < 2 ? _mm_extract_epi64((X).v0, (N) & 0x1)                     \
+                    : _mm_extract_epi64((X).v1, (N) & 0x1);                    \
+  })
+
+#define _mm256_insert_epi32(X, I, N)                                           \
+  __extension__({                                                              \
+    ((N) & 0x7) < 4                                                            \
+      ? _mm256_set_m128i((X).v1, _mm_insert_epi32((X).v0, (I), (N) & 0x3))     \
+      : _mm256_set_m128i(_mm_insert_epi32((X).v1, (I), (N) & 0x3), (X).v0);    \
+  })
+
+#define _mm256_insert_epi16(X, I, N)                                           \
+  ({                                                                           \
```

Comment:
We should use `__extension__` before all of the statement expressions to avoid compiler errors in pedantic mode. Since we're using statement expressions anyway, should we also assign non-constant parameters to variables to prevent them from being evaluated more than once? It would also be good to add `-Wpedantic` to the compilation of these files to test that it works.

Reply:
Yes, I forgot to add __extension__ to some macro, it's fixed now.
Assign non-constant parameters to variables  is also done, -Wpedantic is added to test_avx compilation.

For the build-test failure, one failure is too many locals in generated test_avx.wasm when build with some option, e.g. -O0, it comes from the test case itself, I want to skip some test config, and another failure comes from test-browser, seems the failure case is not directly related to this PR, and I can't reproduce the failure case when run the failure case alone, may need more time to debug this.

Reply:
> too many locals

This error should be possible to fix easily by moving some of the individual function tests into their own functions, and adding a `__attribute__((noinline))` into those test functions. This way the optimizer won't inline these functions into a single massive function that would have way too many locals inline to run in a VM.

Reply:
Try e.g.

```c++
void __attribute__((noinline)) test_arithmetic(void) {
```
in test_avx.cpp or splitting the individual sections there into test_arithmetic_part1, test_arithmetic_part2 etc.

Reply:
yes, I tried to add __attribute__((noinline)) to each test_* function, but this error pop up even when running a single test case, e.g. Ret_M256d_M256d_Tint_5bits(__m256d, _mm256_cmp_pd);  there are about 1600 locals for one specific constant, unrolling 32 times will exceed the limit 50000(1600 * 32 = 51200), split function is easier, but split macro directly doesn't change the locals, so I have to add some function to wrap the splitted macro.